### PR TITLE
[SPARK-39142][SPARK-42235][PYTHON] Add overloads in pandas function stub file

### PR DESCRIPTION
**Pull request by wayneguow**
### What changes were proposed in this pull request?
Add some new overloads in pandas function stub file.

### Why are the changes needed?

Avoiding static type checkers' s warning in some cases without setting `functionType`, and it is preferred to specify type hints for the pandas UDF instead of specifying pandas UDF type via `functionType` which will be deprecated in the future releases.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Pass GA.

### CODE REVIEW

**Comment by wayneguow** [comment]
cc @zero323 and @HyukjinKwon

File: python/pyspark/sql/pandas/functions.pyi [hunk 1]
```diff
@@ -46,13 +46,21 @@ class PandasUDFType:
...
+) -> UserDefinedFunctionLike: ...
 @overload
 def pandas_udf(
     f: PandasScalarToScalarFunction,
     returnType: Union[AtomicDataTypeOrString, ArrayType],
     functionType: PandasScalarUDFType,
 ) -> UserDefinedFunctionLike: ...
 @overload
+def pandas_udf(
+    f: Union[AtomicDataTypeOrString, ArrayType]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/functions.pyi:61]: While we're here, let's just add all data type instead of only Atomic type.

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise. Thanks for fixing this. cc @zero323 too FYI

File: python/pyspark/sql/pandas/functions.pyi (hunk 1 above)
Comment (wayneguow) [reply python/pyspark/sql/pandas/functions.pyi]: Add some other overloads which are same to these two to cover all situations.  (with some ignores to pass `mypy` checks like those existing ignores)

File: python/pyspark/sql/pandas/functions.pyi [hunk 2]
```diff
@@ -46,25 +46,33 @@ class PandasUDFType:
     GROUPED_MAP: PandasGroupedMapUDFType
     GROUPED_AGG: PandasGroupedAggUDFType
 
+@overload
+def pandas_udf(f: PandasScalarToScalarFunction, returnType: Union[AtomicDataTypeOrString, ArrayType]) -> UserDefinedFunctionLike: ...  # type: ignore[misc]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/functions.pyi:50]: Can we add struct type and map types too (instead of Atomic types and array types)?

Comment (wayneguow) [reply python/pyspark/sql/pandas/functions.pyi]: Added `MapType`  for `PandasScalarUDFType` only (cause that `PandasScalarToStructFunction` part exists, no need to add `StructType`), added `MapType, StructType` for `PandasScalarIterUDFType`, added `MapType` for `PandasGroupedAggUDFType`(`StructType` is not supported), and for `PandasGroupedMapFunction`, no need to change.

**Comment by dennisobrien** [comment]
Is there a way to re-open this MR?  It would be nice to have `mypy` and `pyright` not report errors on code using `pandas_udf`.  Is there something I can do to help revive this PR?

**Comment by pontusvision** [comment]
Please reopen this; I'm also experiencing issues with pylance.

**Comment by bjtho08** [comment]
@wayneguow @HyukjinKwon please reopen and merge this PR.

**Comment by AkiraVoid** [comment]
Until now this problem still remains.

**Pull request closed without merging**